spring-boot-starter/starter-security/src/test/java/org/camunda/bpm/spring/boot/starter/security/oauth2/impl/CamundaIdentityProviderIT.java

```diff
@@ -0,0 +1,181 @@
+  }
+
+  @Test
+  public void testIdentityProviderForUsersWithSpringSecurity() throws Exception {
+    // given spring security context
+    ResultCaptor<UserQuery> resultCaptor = new ResultCaptor<>();
+    doAnswer(resultCaptor).when(spiedIdentityProvider).createUserQuery();
+    OAuth2AuthenticationToken authenticationToken = createToken(AUTHORIZED_USER);
+    createAuthorizedClient(authenticationToken, registrations, authorizedClientService);
+
+    // when calling webapp api
+    mockMvc.perform(MockMvcRequestBuilders.get(baseUrl + "/camunda/api/engine/engine/default/user")
+            .accept(MediaType.APPLICATION_JSON)
+            .with(authentication(authenticationToken)))
+        .andDo(MockMvcResultHandlers.print())
+        .andExpect(MockMvcResultMatchers.status().isOk());
+
+    // then identity provider handles oauth2 authentication
+    verify(spiedIdentityProvider, atLeastOnce()).createUserQuery();
+    UserQuery userQueryResult = resultCaptor.result;
+    assertThat(userQueryResult).isInstanceOf(OAuth2IdentityProvider.OAuth2UserQuery.class);
+    assertThat(userQueryResult).isNotNull();
+    assertThat(((OAuth2IdentityProvider.OAuth2UserQuery) userQueryResult).getId()).isEqualTo(AUTHORIZED_USER);
+  }
+
+  @Test
+  public void testIdentityProviderForGroupsWithSpringSecurity() throws Exception {
+    // given spring security context
+    ResultCaptor<GroupQuery> resultCaptor = new ResultCaptor<>();
+    doAnswer(resultCaptor).when(spiedIdentityProvider).createGroupQuery();
+    OAuth2AuthenticationToken authenticationToken = createToken(AUTHORIZED_USER);
+    createAuthorizedClient(authenticationToken, registrations, authorizedClientService);
+
+    // when calling webapp api
+    mockMvc.perform(MockMvcRequestBuilders.get(baseUrl + "/camunda/api/engine/engine/default/user")
+            .accept(MediaType.APPLICATION_JSON)
+            .with(authentication(authenticationToken)))
+        .andDo(MockMvcResultHandlers.print())
+        .andExpect(MockMvcResultMatchers.status().isOk());
+
+    // then identity provider handles oauth2 authentication
+    verify(spiedIdentityProvider, atLeastOnce()).createGroupQuery();
+    GroupQuery groupQueryResult = resultCaptor.result;
+    assertThat(groupQueryResult).isInstanceOf(OAuth2IdentityProvider.OAuth2GroupQuery.class);
+    assertThat(groupQueryResult).isNotNull();
+    assertThat(((OAuth2IdentityProvider.OAuth2GroupQuery) groupQueryResult).getUserId()).isEqualTo(AUTHORIZED_USER);
+  }
+
+  private static void mockIdentityProviderFactory() {
+    // mocks methods of all instances of OAuth2IdentityProviderFactory so instead of always
+    // returning a new instance of the identity provider, it always returns our spy object
+    mockConstruction(OAuth2IdentityProviderFactory.class, (mock, context) -> {
+      doReturn(spiedIdentityProvider).when(mock).openSession();
+      doReturn(WritableIdentityProvider.class).when(mock).getSessionType();
+    });
```

Comment:
Explanation: this method mocks the [openSession()](https://github.com/camunda/camunda-bpm-platform/blob/14a704b522e0cc5c0fada7225a76f3524448120c/spring-boot-starter/starter-security/src/main/java/org/camunda/bpm/spring/boot/starter/security/oauth2/impl/OAuth2IdentityProviderFactory.java#L31) method on **any instance** of `OAuth2IdentityProviderFactory`, so instead of creating a new provider each time, we always return our spy object.